### ISSUE TITLE
getting "mkdir: cannot create directory '/bitnami/mariadb': Permission denied"

### ISSUE THREAD

**Issue by richb201** [bitnami/bitnami-docker-mariadb#193]
I am trying to install a bitnami LAMP image and am following their instructs. They say to run in a compose-docker.yml. 
```yaml
Version: '2'

services:
  apache:
    image: bitnami/apache:latest
    ports:
      - 80:8080
    volumes:
      - ./apache-vhost/myapp.conf:/vhosts/myapp.conf:ro
      - .:/app

  mariadb:
    image: bitnami/mariadb:latest
    environment:
      - MARIADB_ROOT_PASSWORD=cb37trwso*
    volumes:
      - ./mariadb-data:/bitnami

  php-fpm:
    image: bitnami/php-fpm:latest
    volumes:
      - .:/app
```
I have also added a codeigniter  image. 
```yaml
codeigniter:
    image: 'bitnami/codeigniter:3'
    ports:
      - '8000:8000'
    volumes:
      - '.:/app'
    depends_on:
      - mariadb
```
The load of mariaDB is failing with the following error in the log:
```
NFO  ==> ** Starting MariaDB setup **
INFO  ==> Validating settings in MYSQL_*/MARIADB_* env vars..
INFO  ==> Initializing mariadb database...
mkdir: cannot create directory '/bitnami/mariadb': Permission denied
INFO  ==> Stopping mariadb...
```
I am not sure, but I wonder if the directory `/bitnami/mariadb` has permissions that I can change? I tried changing permissions on `/myapp/` to 777 but that didn't seem to work. What am I doing wrong?

**Comment by carrodher** [comment]
Is there any reason to persist the whole `/bitnami` directory instead of `/bitnami/mariadb`?
```yaml
    volumes:
      - 'mariadb_data:/bitnami/mariadb'
```
In your case, you should configure the proper permissions to your local directory `mariadb_data`.
I am not sure if you are using a full `docker-compose.yml` or there are some missing sections, like the volumes one at the end of the file:
```
volumes:
  mariadb_data:
    driver: local
```
Check https://github.com/bitnami/bitnami-docker-mariadb/blob/master/docker-compose.yml

**Comment by richb201** [comment]
Thanks. I am using the bitnami document build a AMP container (https://docs.bitnami.com/containers/how-to/create-amp-environment-containers/). I modified it slightly to add CI and also the password. I am just trying to get my FIRST container up and going. I have copied the actual yml below. I have no idea of why the "WHOLE" directory. I just want to try out getting a container up. 

version: '2'

services:

  apache:
    image: bitnami/apache:latest
    ports:
      - 80:8080
    volumes:
      - ./apache-vhost/myapp.conf:/vhosts/myapp.conf:ro
      - .:/app

  mariadb:
    image: bitnami/mariadb:latest
    environment:
      - MARIADB_ROOT_PASSWORD=abcdefg
    volumes:
      - ./mariadb-data:/bitnami

  php-fpm:
    image: bitnami/php-fpm:latest
    volumes:
      - .:/app

  codeigniter:
    image: 'bitnami/codeigniter:3'
    ports:
      - '8000:8000'
    volumes:
      - '.:/app'
    depends_on:
      - mariadb

**Comment by carrodher** [comment]
I see, there is a typo in the guide you are following, thanks for reporting it. I just created an internal task to fix it.

There are several ways to declare volumes on the fly as part of the service declaration, in this case, it seems we are mixing two concepts:
- On one hand, you can use a local directory (if it doesn't exist, it is created by the Docker Engine)
- On the other hand, you can use named volumes

Using the first approach (the one documented in the guide), you need to ensure that the local folder (`mariadb-data`) has the correct permissions, you can do it just assigning write permissions to `others`:
```bash
sudo chmod o+w mariadb-data
```
After that, you should be able to run
```bash
docker-compose up -d
```
without any problem.
If the `mariadb-data` doesn't exist because is the first `docker-compose up` execution, you can create it and assign the proper permissions:
```bash
mkdir mariadb-data && sudo chmod o+w mariadb-data
```

Using the second approach, you don't need to create the folder or set the proper permissions because the volume is created using the Docker Engine driver as an object instead of a real directory in your filesystem (I think this is the best approach for databases), you only need to edit the  _docker-compose.yml_, adding the following lines to the end
```yaml
 volumes:
   mariadb-data:
     driver: local
```
and changing the local path in the mariadb service
```diff
   mariadb:
     image: bitnami/mariadb:latest
     environment:
       - MARIADB_ROOT_PASSWORD=cb37trwso*
     volumes:
-      - ./mariadb-data:/bitnami
+      - mariadb-data:/bitnami
```

**TL;DR**
There are 2 options to solve the issue:
**Option 1:**
- `mkdir mariadb-data && sudo chmod o+w mariadb-data`
- `docker-compose up`

**Option 2:**
- Replace `./mariadb-data:/bitnami` by `mariadb-data:/bitnami` (removing `./`) in the _docker-compose.yml_ and add the following lines at the end of the file
```yaml
 volumes:
   mariadb-data:
     driver: local
```
- `docker-compose up`

**Comment by carrodher** [comment]
That is because you need to remove a space from the `volumes` code block, it should be at the first level:
```diff
- volumes:
-   mariadb-data:
-     driver: local
+volumes:
+  mariadb-data:
+    driver: local
```
once fixed the typo, you will find other errors:
```
mariadb_1      | 2019-08-26 16:57:51 9 [Warning] Access denied for user 'root'@'172.21.0.5' (using password: NO)
codeigniter_1  | mysql-c ERROR [canConnect] Connection with 'root' user is unsuccessful
```
that is because you configured MariaDB with a password (`MARIADB_ROOT_PASSWORD=abcdefg`) so the CodeIgniter container needs to know this password to connect with the database, so you need to configure the CodeIgniter container via an environment variable.
Other option, if it is not a production application is allow connections without password, on this way you don't need to configure CodeIgniter, you can do it by replacing the password env. variable by:
```diff
- - MARIADB_ROOT_PASSWORD=abcdefg
+ - ALLOW_EMPTY_PASSWORD=yes
```

Using a _docker-compose.yml_ with the above changes (spaces in the volume and `ALLOW_EMPTY_PASSWORD` env. variable) everything work as expected:
```
$ docker-compose up
mariadb_1      | 
mariadb_1      | Welcome to the Bitnami mariadb container
mariadb_1      | Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-mariadb
mariadb_1      | Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-mariadb/issues
mariadb_1      | Send us your feedback at containers@bitnami.com
mariadb_1      | 
mariadb_1      | INFO  ==> ** Starting MariaDB setup **
apache_1       | 
apache_1       | Welcome to the Bitnami apache container
apache_1       | Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-apache
apache_1       | Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-apache/issues
apache_1       | 
mariadb_1      | INFO  ==> Validating settings in MYSQL_*/MARIADB_* env vars..
apache_1       | INFO  ==> Found vhost definitions in /vhosts. Copying them to /opt/bitnami/apache/conf/vhosts
mariadb_1      | WARN  ==> You set the environment variable ALLOW_EMPTY_PASSWORD=yes. For safety reasons, do not use this flag in a production environment.
apache_1       | INFO  ==> Mounting application files from /app...
mariadb_1      | INFO  ==> Initializing mariadb database...
codeigniter_1  | 
codeigniter_1  | Welcome to the Bitnami codeigniter container
codeigniter_1  | Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-codeigniter
codeigniter_1  | Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-codeigniter/issues
codeigniter_1  | 
codeigniter_1  | Starting application ...
codeigniter_1  | Creating example Codeigniter application
mariadb_1      | INFO  ==> Stopping mariadb...
apache_1       | nami    INFO  Initializing apache
apache_1       | nami    INFO  apache successfully initialized
apache_1       | INFO  ==> Starting apache... 
apache_1       | [Mon Aug 26 17:00:52.159715 2019] [ssl:warn] [pid 38] AH01909: www.example.com:8443:0 server certificate does NOT include an ID which matches the server name
apache_1       | [Mon Aug 26 17:00:52.169584 2019] [ssl:warn] [pid 38] AH01909: www.example.com:8443:0 server certificate does NOT include an ID which matches the server name
apache_1       | [Mon Aug 26 17:00:52.172144 2019] [mpm_prefork:notice] [pid 38] AH00163: Apache/2.4.41 (Unix) OpenSSL/1.1.0k configured -- resuming normal operations
apache_1       | [Mon Aug 26 17:00:52.172496 2019] [core:notice] [pid 38] AH00094: Command line: 'httpd -f /opt/bitnami/apache/conf/httpd.conf -D FOREGROUND'
mariadb_1      | INFO  ==> ** MariaDB setup finished! **
mariadb_1      | 
mariadb_1      | INFO  ==> ** Starting MariaDB **
mariadb_1      | 2019-08-26 17:00:53 0 [Note] /opt/bitnami/mariadb/sbin/mysqld (mysqld 10.3.17-MariaDB) starting as process 1 ...
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Using Linux native AIO
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Uses event mutexes
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Compressed tables use zlib 1.2.8
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Number of pools: 1
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Using SSE2 crc32 instructions
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Initializing buffer pool, total size = 128M, instances = 1, chunk size = 128M
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Completed initialization of buffer pool
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Creating shared tablespace for temporary tables
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Waiting for purge to start
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: 10.3.17 started; log sequence number 1630824; transaction id 21
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Loading buffer pool(s) from /bitnami/mariadb/data/ib_buffer_pool
mariadb_1      | 2019-08-26 17:00:53 0 [Note] InnoDB: Buffer pool(s) load completed at 190826 17:00:53
mariadb_1      | 2019-08-26 17:00:53 0 [Note] Plugin 'FEEDBACK' is disabled.
mariadb_1      | 2019-08-26 17:00:53 0 [Note] Server socket created on IP: '0.0.0.0'.
mariadb_1      | 2019-08-26 17:00:53 0 [Warning] 'proxies_priv' entry '@% root@83427e84691c' ignored in --skip-name-resolve mode.
mariadb_1      | 2019-08-26 17:00:53 0 [Note] Reading of all Master_info entries succeeded
mariadb_1      | 2019-08-26 17:00:53 0 [Note] Added new Master_info '' to hash table
mariadb_1      | 2019-08-26 17:00:53 0 [Note] /opt/bitnami/mariadb/sbin/mysqld: ready for connections.
mariadb_1      | Version: '10.3.17-MariaDB'  socket: '/opt/bitnami/mariadb/tmp/mysql.sock'  port: 3306  Source distribution
codeigniter_1  | mysql-c INFO  Trying to connect to MySQL server
codeigniter_1  | mysql-c INFO  Found MySQL server listening at mariadb:3306
codeigniter_1  | mysql-c INFO  MySQL server listening and working at mariadb:3306
codeigniter_1  | codeign INFO  The Codeigniter project myapp was created succesfully!
codeigniter_1  | Codeigniter app created
```
To obtain this result I used your _docker-compose.yml_ adding the mentioned changes.

**Comment by carrodher** [comment]
Using the `ALLOW_EMPTY_PASSWORD` everything should work without issues nor new changes, if you need to use the password, yes, it should be included at deploy time because the codeigniter container needs a connection with the mariadb one, you're adding it as a dependency:
```yaml
    depends_on:
      - mariadb
```

About resources to learn, Docker has an [official Get Started guide](https://docs.docker.com/get-started/) and also there are some interactive resources like [Play with Docker](https://training.play-with-docker.com/)

**Comment by jmpolvera** [comment]
I made this in the host directory and worked:

sudo chown -R 1001:1001 <directory>


https://techoverflow.net/2018/12/15/solving-bitnami-docker-redmine-cannot-create-directory-bitnami-mariadb-permission-denied/
